src/triangle.rs

```diff
@@ -0,0 +1,647 @@
+// Copyright 2024 the Kurbo Authors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! Triangle shape
+use crate::{Ellipse, PathEl, Point, Rect, Shape, Size, Vec2};
+
+use core::cmp::*;
+use core::f64::consts::FRAC_PI_4;
+use core::ops::{Add, Sub};
+
+#[cfg(not(feature = "std"))]
+use crate::common::FloatFuncs;
+
+/// Triangle
+//     A
+//     *
+//    / \
+//   /   \
+//  *-----*
+//  B     C
+#[derive(Clone, Copy, PartialEq, Debug)]
+#[cfg_attr(feature = "schemars", derive(schemars::JsonSchema))]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct Triangle {
+    /// vertex a
+    pub a: Point,
+    /// vertex b
+    pub b: Point,
+    /// vertex c
+    pub c: Point,
+}
+
+impl Default for Triangle {
+    fn default() -> Self {
+        Self::EQUILATERAL
+    }
+}
+
+macro_rules! most_ordering {
+    ($x: ident, $by: block) => {
+        *$x.as_array().iter().max_by($by).unwrap()
+    };
+}
+
+impl Triangle {
+    /// The empty [`Triangle`] at the origin
+    pub const ZERO: Self = Self::from_coords((0., 0.), (0., 0.), (0., 0.));
+
+    /// empty [`Triangle`] at (1.0, 1.0)
+    pub const ONE: Self = Self::from_coords((1.0, 1.0), (1.0, 1.0), (1.0, 1.0));
+
+    /// equilateral [`Triangle`] identity
+    pub const EQUILATERAL: Self = Self::from_coords((3.0, 6.0), (0.0, 0.0), (6.0, 0.0));
+
+    /// A new [`Triangle`] from three vertices ([`Points`])
+    #[inline]
+    pub fn new(a: impl Into<Point>, b: impl Into<Point>, c: impl Into<Point>) -> Self {
+        Self {
+            a: a.into(),
+            b: b.into(),
+            c: c.into(),
+        }
+    }
+
+    /// A new [`Triangle`] from three float vertex coordinates
+    #[inline]
+    pub const fn from_coords(a: (f64, f64), b: (f64, f64), c: (f64, f64)) -> Self {
+        Self {
+            a: Point::new(a.0, a.1),
+            b: Point::new(b.0, b.1),
+            c: Point::new(c.0, c.1),
+        }
+    }
+
+    /// A new [`Triangle`] from centroid and sizes ([`Size`]) - offset of each vertex from the centroid
+    #[inline]
+    pub fn from_centroid_sizes(centroid: impl Into<Point>, sizes: [impl Into<Size>; 3]) -> Self {
+        let centroid = centroid.into();
+
+        let mut points: [Point; 3] = Default::default();
+        for (i, size) in sizes.into_iter().enumerate() {
+            points[i] = centroid + size.into().to_vec2();
+        }
+
+        Self::new(points[0], points[1], points[2])
+    }
+
+    /// A new [`Triangle`] with vertex distances of `distances` from the `centroid`
+    ///
+    /// NOTE: the new [`Triangle`] does not keep `centroid`
+    #[inline]
+    pub fn from_centroid_distances(centroid: impl Into<Point>, distances: [f64; 3]) -> Self {
+        let centroid = centroid.into();
+
+        Self::new(
+            centroid + (0.0, distances[0]),
+            centroid + distances[1] * Vec2::from_angle(5.0 * FRAC_PI_4),
+            centroid + distances[2] * Vec2::from_angle(7.0 * FRAC_PI_4),
+        )
+    }
```

Comment:
maybe a transition from `from*` to `with*` like in the [builder pattern](https://en.wikipedia.org/wiki/Builder_pattern) is better. although im hesitant to do this here because this would require a change for the entire codebase so maybe should be another pr?

Reply:
we also wouldnt have to include the `centroid` param this way since users could chain with `with_centroid()`